tools/build_langserver/langserver/rename.go

```diff
@@ -0,0 +1,164 @@
+package langserver
+
+import (
+	"context"
+	"core"
+	"encoding/json"
+	"parse/asp"
+	"tools/build_langserver/lsp"
+
+	"github.com/sourcegraph/jsonrpc2"
+)
+
+const renameMethod = "textDocument/rename"
+
+func (h *LsHandler) handleRename(ctx context.Context, req *jsonrpc2.Request) (result interface{}, err error) {
+	if req.Params == nil {
+		return nil, nil
+	}
+
+	var params lsp.RenameParams
+	if err := json.Unmarshal(*req.Params, &params); err != nil {
+		return nil, err
+	}
+
+	documentURI, err := getURIAndHandleErrors(params.TextDocument.URI, renameMethod)
+	if err != nil {
+		return nil, err
+	}
+
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	edits, err := h.getRenameEdits(ctx, params.NewName, documentURI, params.Position)
+
+	return edits, nil
+}
+
+func (h *LsHandler) getRenameEdits(ctx context.Context, newName string,
+	uri lsp.DocumentURI, pos lsp.Position) (*lsp.WorkspaceEdit, error) {
+
+	// Get the document stored
+	doc, ok := h.workspace.documents[uri]
+	if !ok {
+		log.Info("document %s is not opened", uri)
+		return nil, nil
+	}
+
+	def, err := h.analyzer.BuildDefsFromPos(ctx, uri, pos)
+	if def == nil || err != nil || !isPosAtNameArg(def, pos) {
+		return nil, err
+	}
+
+	coreLabel, err := getCoreBuildLabel(def, uri)
+	if err != nil {
+		return nil, err
+	}
+	renamingLabel, err := h.analyzer.BuildLabelFromCoreBuildLabel(ctx, coreLabel)
+	if err != nil {
+		return nil, err
+	}
+
+	revDeps, err := h.analyzer.RevDepsFromCoreBuildLabel(coreLabel, uri)
+	if err != nil {
+		log.Info("error occurred computing the reverse dependency of %")
```

Comment:
this doesn't look right (Github recognises that something is up, didn't know it did that!)